controllers/control/cassandratask_controller.go

```diff
@@ -0,0 +1,593 @@
+					logger.V(1).Info("this job isn't allowed to run due to ConcurrencyPolicy restrictions", "activeTasks", len(activeTasks))
+					return ctrl.Result{Requeue: true}, nil // TODO Add some sane time here
+				}
+			}
+		}
+
+		// Link this resource to the Datacenter and copy labels from it
+		if err := controllerutil.SetOwnerReference(&dc, &cassTask, r.Scheme); err != nil {
+			logger.Error(err, "unable to set ownerReference to the task", "Datacenter", cassTask.Spec.Datacenter, "CassandraTask", req.NamespacedName)
+			return ctrl.Result{}, err
+		}
+
+		utils.MergeMap(cassTask.Labels, dc.GetDatacenterLabels())
+		oplabels.AddOperatorLabels(cassTask.GetLabels(), &dc)
+
+		// Starting the run, set the Active label so we can quickly fetch the active ones
+		cassTask.GetLabels()[taskStatusLabel] = activeTaskLabelValue
+
+		if err := r.Client.Update(ctx, &cassTask); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		cassTask.Status.StartTime = &timeNow
+		cassTask.Status.Active = 1 // We don't have concurrency inside a task at the moment
+	}
+
+	var res ctrl.Result
+	completedCount := int32(0)
+
+	// TODO We only support a single job at this stage (helps dev work)
+	if len(cassTask.Spec.Jobs) > 1 {
+		return ctrl.Result{}, fmt.Errorf("only a single job can be defined in this version of cass-operator")
+	}
+
+	// TODO We need to verify that the cluster is actually healthy before we run these..
+	taskId := string(cassTask.UID)
+
+	var err error
+	for _, job := range cassTask.Spec.Jobs {
+		switch job.Command {
+		case "rebuild":
+			res, err = r.reconcileEveryPodTask(ctx, taskId, &dc, rebuild(job.Arguments))
+		case "cleanup":
+			res, err = r.reconcileEveryPodTask(ctx, taskId, &dc, cleanup())
+		case "decommission":
+			// res, err = r.reconcileEveryPodTask(ctx, taskId, &dc, decommission())
+		default:
+			err = fmt.Errorf("unknown job command: %s", job.Command)
+			return ctrl.Result{}, err
+		}
+
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if res.RequeueAfter > 0 {
+			// This job isn't complete yet or there's an error, do not continue
+			break
+		}
+		completedCount++
+	}
+
+	if res.RequeueAfter == 0 && !res.Requeue {
+		// Job has been completed
+		cassTask.GetLabels()[taskStatusLabel] = completedTaskLabelValue
+		if err = r.Client.Update(ctx, &cassTask); err != nil {
+			return res, err
+		}
+
+		r.cleanupJobAnnotations(ctx, &dc, taskId)
+
+		cassTask.Status.Active = 0
+		cassTask.Status.CompletionTime = &timeNow
+
+		// Requeue for deletion later
+		deletionTime := calculateDeletionTime()
+		nextRunTime := deletionTime.Sub(timeNow.Time)
+		res.RequeueAfter = nextRunTime
+		logger.V(1).Info("This task has completed, scheduling for deletion in " + nextRunTime.String())
+	}
+
+	if cassTask.Status.Succeeded != completedCount {
+		cassTask.Status.Succeeded = completedCount
+	}
+
+	// TODO Add conditions also
+	if err = r.Client.Status().Update(ctx, &cassTask); err != nil {
+		return ctrl.Result{}, err
```

Comment:
At one point I tested with a bad management-api image which resulted in an error when `taskConfig.AsyncFunc` is called. The failure was due to the RPC endpoint not being properly exposed. The errors happens against the first pod, subsequent pods are never tried. At some point the job should be marked failed. k8ssandra-operator ends up in a requeue loop waiting for the  job to complete, but it never does.

Reply:
It seems my reply never got here, odd.. well, the process catches failed jobs and exits with them correctly and cass-operator should kill pods that are not healthy. However, you hit an edge case where management-api is just simply not working correctly but cass-operator is not killing it.